### ISSUE TITLE
Some css styles are not created in production but work in development

### ISSUE THREAD

**Issue by jjlauer** [JedWatson/react-select#3309]
Hello,

We are hitting an odd css/emotion issue in production, while the component works great in development. We've tried this on v2.0.0-v.2.2.0 and its the same issue with all of them.

The dynamic css rules for the outer divs never get created in the emotion stylesheet.  They are, however, set on the divs in the DOM. The rest of the dynamic css rules (e.g. on the options when you click on the control) do get created in the emotion stylesheet. This only happens on a few pages in our app, while other parts work as expected. Here is a screenshot of the the DOM when the page loads.

![screenshot from 2019-01-01 19-22-05](https://user-images.githubusercontent.com/585730/50577664-cfc09280-0dfa-11e9-8930-36b2191a0c7c.png)

You'll notice that the select outermost div was assigned a classname of `css-10nd86i`, but that css rule does not actually exist in the DOM. Using the javascript console to dump out the emotion stylesheet, you'll notice in this screenshot that the rules are totally empty in the stylesheet.

![screenshot from 2019-01-01 19-23-24](https://user-images.githubusercontent.com/585730/50577682-452c6300-0dfb-11e9-85cf-f5996487cc9d.png)

Now if you click on the select control to open it, the css rules for the options list all get created:

![screenshot from 2019-01-01 19-43-12](https://user-images.githubusercontent.com/585730/50577773-89b8fe00-0dfd-11e9-86dc-3c46244fbc45.png)

So the options get styled correctly, but the primary select component does not. I've tried various versions of emotion, react-select, react, etc., but the issue still persists.  We've had to swap out this component in production for the time being, but would like to figure this out.  Any help would be greatly appreciated.

**Comment by gwyneplaine** [comment]
@jjlauer are there any key similarities between the pages where this happens? Any information you could provide us to narrow down a reproducible case for this would be helpful; and if possible a [codesandbox](https://codesandbox.io/s/new) reproduction.

This usually happens when you're trying to instantiate emotion across contexts, like in an iframe or through server side rendering for example.

**Comment by jjlauer** [comment]
Made some headway. On the page in our app that works, it's a very complicated page and the user would only see the Select component if they were already on the page for a few seconds (e.g. javascript all loaded, before Select ever rendered).

On the page this issue occurs the Select component is rendered immediately when the browser loads the page. If I delay the Select from being rendered for even just 100ms (via setTimeout) then it renders correctly.  Any ideas on what could be causing something like that and why even just 100ms delay in React rendering the Select component would fix it?

**Comment by ryanrombough** [comment]
@jjlauer Did you ever figure this out? I'm having the same issue. The selector is properly styled in a development environment, but in a production build none of the emotion stylesheets get injected into the head.

**Comment by jjlauer** [comment]
@ryanrombough Unfortunately no. Only workaround is what i mentioned above -- you have to delay rendering of the widget so it doesn't occur on the first react render().

**Comment by ryanrombough** [comment]
@jjlauer I just tried implementing a 500ms delay on rendering but it did not solve the issue for me. Guess I'll keep digging. Thanks!

**Comment by ryanrombough** [comment]
I have narrowed the cause of my issue down to the minification step in my Webpack v4 build. Here's the [stack overflow question]( https://stackoverflow.com/questions/55505056/why-would-webpack-4-minification-prevent-styles-for-react-select-component).

**Comment by jjlauer** [comment]
Hmm.. did turning off minimize in webpack fix your production issue?

I also had issues w/ the css not being injected into the header, but adding
a delay (of any use of react-select) on the initial react render fixed it.
It feels like a race condition between Emotion fully initializing before
react-select uses it.

**Comment by ryanrombough** [comment]
Yea, but that's not really a solution for us because our bundle is too large when not minified. I tried implementing the delay you suggested but could not get it to work.

**Comment by ChayanJana1233** [comment]
Hi @jjlauer I am facing the same CSS not being applied in Production. If you can kindly help me with the code snippet for setting the delay in rendering the react-select it would be extremely helpful!!
Thank you.

**Comment by hemedani** [comment]
I have the same issues

**Comment by jwpihlgren** [comment]
I experience the same issue.

**Comment by bladey** [comment]
Thanks for letting us know @NearHuscarl!

**Comment by kunambi** [comment]
@NearHuscarl's solution didn't solve my problem. Have used `classNamePrefix` for a while, and the issue remains the same.

@jjlauer how did you implement the delay?

@bladey is there a way to "force" the CSS to be loaded programatically?

Getting desperate here 😅

**Comment by jjlauer** [comment]
Here is how I implemented a very simple delay to rendering.  Here's the main react component (e.g. top-level entrypoint).  I originally was rendering the `ReportsPanel` component, but developed a simple wrapper component i called `DelayedRender` that is now responsible for rendering it instead.

```
class ReportsPage extends Component {

  constructor() {
    super();
  }

  render() {
    const { dispatch, query, user, accounts } = this.props;

    return (
      <DelayedRender>
        <ReportsPanel
          dispatch={dispatch}
          query={query}
          user={user}
          accounts={accounts}
          active={true}
        />
      </DelayedRender>
    );
  }
}
```
Here is DelayedRender:

```
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class DelayedRender extends PureComponent {

  constructor() {
    super();
    
    this.state = {
      render: false
    };
  }
  
  componentDidMount() {
    const { time } = this.props;
    
    setTimeout(() => this.setState({ render: true }), (time || 400));
  }
  
  render() {
    const { children } = this.props;
    const { render } = this.state;
    
    if (!render) {
      return null;
    }
    
    return children;
  }
}
```

**Comment by wegry** [comment]
This seems like it's still present in react-select@4.2.1.

**Comment by rushilsrivastava** [comment]
This problem has reappeared for us in 4.2.1

**Comment by wegry** [comment]
@rushilsrivastava are you folks using `react-windowed-select` as well as `react-select`? My team has a hunch it may be the former causing issues.

Confusingly, downgrading just `react-windowed-select` to use `react-select` 3 seems to resolve the issue.

**Comment by rushilsrivastava** [comment]
@wegry We aren't using `react-windowed-select`, still having a hard time pinpointing what package/code change caused this since downgrading to 4.1.0 doesn't seem to work for us either. We will give `react-select` v3 a shot though and report back here.

**Comment by ebonow** [comment]
Greetings @wegry , @rushilsrivastava , @jjlauer , or anyone else, 

Would any of you be willing to share your package.json to help us recreate the issue?

**Comment by wegry** [comment]
@ebonow 

```json
{
  "browserslist": [
    "last 2 chrome versions",
    "last 2 firefox versions",
    "last 1 safari versions"
  ],
  "dependencies": {
    "@fortawesome/fontawesome-svg-core": "^1.2.30",
    "@fortawesome/free-solid-svg-icons": "^5.14.0",
    "@fortawesome/react-fontawesome": "^0.1.11",
    "@rjsf/bootstrap-4": "2.4.0",
    "@rjsf/core": "2.4.0",
    "@types/color-hash": "^1.0.0",
    "availity-reactstrap-validation": "^2.6.1",
    "axios": "^0.19.2",
    "bootstrap": "^4.6.0",
    "buffer": "^5.6.0",
    "classnames": "^2.2.6",
    "color-hash": "^1.0.3",
    "color2k": "^1.1.0",
    "core-js": "^3.9.0",
    "core-util-is": "^1.0.2",
    "css-loader": "^5.0.2",
    "dayjs": "^1.10.3",
    "events": "^3.2.0",
    "fp-ts": "^2.9.3",
    "immer": "^8.0.1",
    "imports-loader": "^2.0.0",
    "io-ts": "^2.2.13",
    "io-ts-reporters": "^1.2.2",
    "js-cookie": "rc",
    "json-schema-from-path": "^2.0.1",
    "json-schema-traverse": "^0.4.1",
    "localforage": "^1.9.0",
    "lodash": "^4.17.19",
    "lodash-es": "^4.17.15",
    "memoize-one": "^5.1.1",
    "mobx": "^6.1.7",
    "mobx-react": "^7.1.0",
    "mobx-react-lite": "^3.2.0",
    "moment": "^2.29.1",
    "moment-timezone": "^0.5.32",
    "monaco-editor": "^0.22.3",
    "parse-data-url": "^3.0.0",
    "path-browserify": "^1.0.1",
    "process": "^0.11.10",
    "prop-types": "^15.6.2",
    "punycode": "^2.1.1",
    "query-string": "^6.8.2",
    "querystring-es3": "^0.2.1",
    "rc-slider": "9.2.4",
    "react": "^16.14.0",
    "react-big-calendar": "^0.19.2",
    "react-bootstrap": "^1.5.0",
    "react-csv": "^2.0.3",
    "react-datetime": "^3.0.4",
    "react-dom": "^16.14.0",
    "react-event-timeline": "1.6.3",
    "react-json-view": "^1.21.1",
    "react-modal": "^3.11.2",
    "react-monaco-editor": "^0.42.0",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0",
    "react-select": "^4.1.0",
    "react-spring": "^8.0.27",
    "react-table": "^7.6.3",
    "react-toastify": "^7.0.0",
    "react-twitter-widgets": "^1.9.5",
    "react-use": "^17.1.1",
    "react-windowed-select": "^2.0.2",
    "reactstrap": "^8.6.0",
    "recharts": "^2.0.8",
    "semver": "^7.3.2",
    "stream-browserify": "^3.0.0",
    "string-similarity": "^4.0.4",
    "string_decoder": "^1.3.0",
    "timers-browserify": "^2.0.12",
    "tslib": "^2.0.3",
    "use-immer": "^0.4.2",
    "util": "^0.12.3",
    "uuid": "^8.3.1",
    "xml2js": "^0.4.19"
  },
  "devDependencies": {
    "@babel/core": "^7.13.8",
    "@babel/plugin-proposal-class-properties": "^7.13.0",
    "@babel/plugin-proposal-decorators": "^7.13.5",
    "@babel/preset-env": "^7.13.9",
    "@babel/preset-react": "^7.12.13",
    "@babel/preset-typescript": "^7.13.0",
    "@fortawesome/fontawesome-free": "^5.14.0",
    "@jest/globals": "^26.6.2",
    "@types/classnames": "^2.2.10",
    "@types/css-minimizer-webpack-plugin": "^1",
    "@types/enzyme": "^3.10.8",
    "@types/events": "^3",
    "@types/js-cookie": "^2",
    "@types/json-schema": "^7.0.6",
    "@types/json-schema-traverse": "^0.4.0",
    "@types/lodash-es": "^4.17.3",
    "@types/punycode": "^2",
    "@types/rc-slider": "^8.6.6",
    "@types/react": "^16.14.3",
    "@types/react-csv": "^1.1.1",
    "@types/react-dom": "^16.9.10",
    "@types/react-modal": "^3.10.6",
    "@types/react-router": "^5.1.8",
    "@types/react-router-dom": "^5.1.5",
    "@types/react-select": "^4.0.5",
    "@types/react-table": "^7.0.24",
    "@types/reactstrap": "^8.5.0",
    "@types/semver": "^7.3.1",
    "@types/string-similarity": "^4.0.0",
    "@types/uuid": "^8.3.0",
    "@typescript-eslint/eslint-plugin": "^4.16.1",
    "@typescript-eslint/parser": "^4.16.1",
    "autoprefixer": "^10.2.2",
    "babel-eslint": "^10.1.0",
    "babel-jest": "^26.6.3",
    "babel-loader": "^8.2.2",
    "babel-plugin-jsx-control-statements": "^4.1.0",
    "chromedriver": "^88.0.0",
    "cross-env": "^7.0.3",
    "css-minimizer-webpack-plugin": "^1.2.0",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.6",
    "esbuild-loader": "^2.9.2",
    "eslint": "^7.21.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-plugin-react": "^7.22.0",
    "eslint-plugin-react-hooks": "^4.2.0",
    "eslint-webpack-plugin": "^2.5.2",
    "fork-ts-checker-webpack-plugin": "^6.1.0",
    "geckodriver": "^1.16.2",
    "html-webpack-plugin": "^5.1.0",
    "jest": "^26.6.3",
    "json5": "^2.2.0",
    "mini-css-extract-plugin": "^1.2.1",
    "monaco-editor-webpack-plugin": "^3.0.0",
    "nightwatch": "^1.5.1",
    "node-rest-client": "^3.1.0",
    "postcss": "^8.2.5",
    "postcss-loader": "^5.0.0",
    "prettier": "^2.2.1",
    "react-test-renderer": "^16.14.0",
    "sass": "^1.32.7",
    "sass-loader": "^11.0.1",
    "style-loader": "^2.0.0",
    "typescript": "^4.2.2",
    "webpack": "^5.24.3",
    "webpack-cli": "^4.5.0",
    "webpack-dev-server": "^3.11.2"
  },
  "dependenciesMeta": {
    "core-js": {
      "built": false
    }
  },
  "license": "UNLICENSED"
}
```

**Comment by Methuselah96** [comment]
@wegry Can you look at your `yarn.lock` or `package-lock.json` and see if anything depends on Emotion besides `react-select`?

**Comment by Methuselah96** [comment]
@wegry I looked into your `package.json` and it looks like you have `react-windowed-select@^2.02` installed which references `react-select@3`. Can you try upgrading that to `react-windowed-select@3.0.0` and see if that makes a difference? `react-windowed-select@2` depends on `react-select@3`, but `react-windowed-select@3` depends on `react-select@4`, so that mismatch could be a potential cause of the issue.

**Comment by wegry** [comment]
@Methuselah96 let me post my whole package.json and yarn.lock as a gist. The version above is current after the downgrade. I'll send the version reproing this instead. My mistake.

**Comment by wegry** [comment]
Just so it's clear that I added a gist (not sure if edits notify), https://gist.github.com/wegry/5b2d97db8257b459b95c08c1e3153dd5. @Methuselah96

**Comment by Methuselah96** [comment]
@wegry Yeah, I'm not seeing anything suspicious. At this point the only thing to do is to try to recreate the issue in a public repo so that we can dig into it more. Are you using Webpack? Are there CSS loaders that you're using that could be affecting this?

**Comment by wegry** [comment]
@Methuselah96 I've unfortunately been unable to repro this at a smaller scale than the app it's occuring in.

> Are you using Webpack?

Yeah, webpack@5.26.3. 

I'm thinking in our case, it's because we have two+ chunks that pull in react-select. The way we repro the styles visibly being broken (and not present in the dom) is

1. Prod build
1. cache free refreshing chunk A where we import `react-select`'s default export.
1. Navigating to a page (in a different chunk) where a wrapped component containing three different versions of react-select (one of which is `'react-select/creatable'`) and react-windowed-select is used.

There might be multiple emotion caches being used in our case?

> Are there CSS loaders that you're using that could be affecting this?

I would think CSS loaders wouldn't affect this, especially since react-select doesn't package css, right?

**Comment by erkez** [comment]
I had the issue with the styles not working with production build after some of my components were inserted into a shadow root. What worked for me was to add `CacheProvider` from emotion somewhere in the application root.

Please see https://github.com/JedWatson/react-select/issues/3680#issuecomment-809518245 for more details.

Hope this helps.

**Comment by piotrrussw** [comment]
Any update or solution on this issue? Apart from delay rendering

**Comment by arda-** [comment]
I had this issue with `react-select` 4.3.0, and @NearHuscarl 's solution https://github.com/JedWatson/react-select/issues/3309#issuecomment-678728542 worked for me! 

stack:
- gatsby (3.4.1)
- react (11.4.0)
- emotion (gatsby-plugin-emotion: 6.5.0)
- twin.macro (2.4.1)
- react-select (4.3.0)

issue:
- a route page (`/myRoute`)  in my app contained a `react-select` component
- on _other pages_, **hovering** on gatsby links `<Link to="/myRoute">` caused the styles of unrelated components on the _other pages_ to break. 
   - this only happened on **production builds**. in the inspector, I could see that the links to `/myRoute` caused chunk load errors on hover.
   - likely due to gatsby's behavior of chunk splitting in prod builds, and pre-fetching chunks when hovering links.

solution:
- adding a `classNamePrefix="react-select"` to our `<Select` component made the other pages' pre-fetched loaded css chunks not fail.

**Comment by myorkgitis** [comment]
I was able to come up with a simple workaround, but can only guess as to why it works. 

**TLDR**
I hid a `<Select>` in the `_app.tsx` outside of the `<Component>`.

```
// _app.tsx
function MyApp({Component, pageProps}: AppProps) {
    return (
        <>
        <div>
            // This is the User Context Provider from auth0
            <UserProvider>
                <div className="app-container">
                    <Component {...pageProps} />
                </div>
            </UserProvider>

            // Hide this select from view
            <div style={{display: "none"}}>

                // This select is put here to get the emotion styles rendered in production */}
                <Select
                    instanceId={"rendered-select"}
                    value={{label: "none", value: "none"}}
                    options={[{label: "none", value: "none"}]}
                    onChange={() => null}
                />
            </div>
        </div>
        </>
    )
}

export default MyApp
```

**My guess as to why this works**
I had this issue running `react-select` 4.3.1 on next.js and tried the previously mentioned solutions but they did not work (delayed rendering https://github.com/JedWatson/react-select/issues/3309#issuecomment-714758497 worked occasionally, but not every time). 

Since the delayed rendering worked occasionally, I thought it could be a caching issue (or some other race condition related to the inner workings of react/next. 

I am using `auth0` and wrapping every page in the `UserProvider` from `auth0` in my `_app.tsx`. `UserProvider` then provides context via a `useUser`.

```
// Every page needing app user context uses this wrapper
const PageWrapper = ({children, serializedDsUser} : PageWrapperProps) => {
    // authUser is the auth0 user
    const { user: authUser, isLoading } = useUser()

    // dsUser is the app's user object
    const dsUser = serializedDsUser ? JSON.parse(serializedDsUser) : null

    const context: UserSessionContextValues = {
        authUser: authUser ? authUser : null,
        dsUser,
        isLoggedIn: authUser ? true : false
    }

    return (
        <div>
            {isLoading ?
                <AppLoader isLoading />
                :
                <UserSessionContext.Provider value={context}>
                {children}
                </UserSessionContext.Provider>
            }
        </div>
    )
}

export default PageWrapper
```

Each page does not render until user data returned by `useUser` completes loading. Without spending many more hours tracing this bug down, I'm guessing something having to do with conditionally rendering the pages and subsequently any nested `<Select>` components based on user state and SSR is causing some sort of race condition with the caching (as mentioned by @jjlauer  https://github.com/JedWatson/react-select/issues/3309#issuecomment-479678164).

Hope this saves someone some time.

**Comment by mattvb91** [comment]
None of the above with hiding in _app.tsx or setting a classname prefix works for me on nextjs currently on client route push. Has anyone been able to get this working recently?

**Comment by kunambi** [comment]
> None of the above with hiding in _app.tsx or setting a classname prefix works for me on nextjs currently on client route push. Has anyone been able to get this working recently?
> 
> delayed rendering doesnt work either i cant get the css to show up if im including the select conditionaly based on state

Had the same issue - nothing helped. 

I only managed to make it work by, on dev server I inspected each and every element copied the CSS to my own CSS-files. Make sure that you lock the exact version so a future update doesn't break the layout.

**Comment by mattvb91** [comment]
@kunambi thanks for the tip. Im assuming this just went into your global css? Will give this a try later

**Comment by mattvb91** [comment]
@kunambi how did you turn off the dynamic classnames? just by giving it a classPrefix?

**Comment by kunambi** [comment]
> @kunambi how did you turn off the dynamic classnames? just by giving it a classPrefix?

Yes, I pretty much hardcoded everything to make sure it works all the time. Also locked the version number to `3.1.0`

**Comment by Fley** [comment]
For us the issue occurred when the chrome tab using react-select was left idle for some time (could be hours). When going back to this tab the issue was there.

I implemented the same kind of solution as @kunambi. On my side I chose to bypass the emotion styles by passing my "own internal components" via the `components` property. The components I use are actually the original lib components but where I force my own css classes (see example below) 🤷 

It was a bit painful to do, but since we've release it we don't have the issue anymore.

```jsx
export const MultiValueLabel = (props) => {
  return (
    <components.MultiValueLabel
      {...props}
      innerProps={{
        ...props.innerProps,
        className: "multiValueLabel",
      }}
    />
  );
};
```

**Comment by mattvb91** [comment]
Thanks for the input guys, im afraid I don't have enough deep knowledge / lacking time to implement a fix so ive just had to go down the route of not using next/link and making the app do a hard refresh on the pages im using the component. Maybe its fixed in the next major version.

**Comment by EandrewJones** [comment]
I am currently encountering this in v5.2.2.

**Comment by mattvb91** [comment]
Update for everyone it ended up being this library that was causing the issue for me: https://github.com/google-map-react/google-map-react

There must be some onload event in that library that triggered some css removal. I cant exactly remember what the fix was but I think I needed to use a different import for the component I was using.

The way I figured it out was start a clean project with only react-select and then start adding libraries from my other project until I started getting the same issue again.

**Comment by marcomartinscastanho** [comment]
I'm having the same issue appearing in v3.2.0.
In my project, we were already using the `classNamePrefix` as suggested by @NearHuscarl.

[local | prod]
![image](https://user-images.githubusercontent.com/30021123/171831438-e2887c36-c1b2-4b33-942e-16ba7a16a610.png)

I can see that the class is added to the component in both cases, it just seems like it isn't loaded in production.

**Comment by jvannistelrooy** [comment]
I had a similar problem with the react-select 5.7.0, Next.js 13.3, and server-side rendering. I used the storybook example mentioned in the react-select documentation about classNames here:
https://react-select.com/styles#the-classnames-prop

useMemo in this part of the example doesn't work well with SSR:
`const EmotionCacheProvider = ({ children }: { children: React.ReactNode }) => {
  const cache = React.useMemo(
    () =>
      createCache({
        key: 'with-tailwind',
        insertionPoint: document.querySelector('title')!,
      }),
    []
  );

  return <CacheProvider value={cache}>{children}</CacheProvider>;
};`

Removing the useMemo part fixed the problem for me.

**Comment by mstajic-bosch** [comment]
Hi folks, its been a while since the last comment, any updates on this, I have tried delayed rendering and it does not work. Using in a SPA mode,  using 5.8.0 version
